pkg/scheduler/api/job_info.go

```diff
@@ -485,24 +525,73 @@ func (ji JobInfo) String() string {
 // FitError returns detailed information on why a job's task failed to fit on
 // each available node
 func (ji *JobInfo) FitError() string {
+	sortReasonsHistogram := func(reasons map[string]int) []string {
+		reasonStrings := []string{}
+		for k, v := range reasons {
+			reasonStrings = append(reasonStrings, fmt.Sprintf("%v %v", v, k))
+		}
+		sort.Strings(reasonStrings)
+		return reasonStrings
+	}
+
+	// Stat histogram for all tasks of the job
 	reasons := make(map[string]int)
 	for status, taskMap := range ji.TaskStatusIndex {
 		reasons[status.String()] += len(taskMap)
 	}
 	reasons["minAvailable"] = int(ji.MinAvailable)
+	reasonMsg := fmt.Sprintf("%v, %v", scheduling.PodGroupNotReady, strings.Join(sortReasonsHistogram(reasons), ", "))
 
-	sortReasonsHistogram := func() []string {
-		reasonStrings := []string{}
-		for k, v := range reasons {
-			reasonStrings = append(reasonStrings, fmt.Sprintf("%v %v", v, k))
+	// Stat histogram for pending tasks only
+	reasons = make(map[string]int)
+	for _, task := range ji.Tasks {
+		if task.Status != Pending {
+			continue
 		}
-		sort.Strings(reasonStrings)
-		return reasonStrings
+		reason, _ := ji.TaskSchedulingReason(task.UID)
+		reasons[reason]++
+	}
+	if len(reasons) > 0 {
+		reasonMsg += "; " + fmt.Sprintf("%s: %s", Pending.String(), strings.Join(sortReasonsHistogram(reasons), ", "))
 	}
-	reasonMsg := fmt.Sprintf("%v, %v.", scheduling.PodGroupNotReady, strings.Join(sortReasonsHistogram(), ", "))
 	return reasonMsg
 }
 
+// TaskSchedulingReason get detailed reason and message of the given task
+//   Returns detailed reason and message for tasks based on last scheduling transaction
+func (ji *JobInfo) TaskSchedulingReason(tid TaskID) (reason string, msg string) {
+	taskInfo, exists := ji.Tasks[tid]
+	if !exists {
+		return "", ""
+	}
+
+	// Get detailed scheduling reason based on LastTransaction
+	ctx := taskInfo.GetTransactionContext()
+	if taskInfo.LastTransaction != nil {
+		ctx = *taskInfo.LastTransaction
+	}
+
+	msg = ji.JobFitErrors
+	switch status := ctx.Status; status {
+	case Allocated, Pipelined:
+		// Pod is schedulable
+		msg = fmt.Sprintf("Pod %s/%s can possibly be assigned to %s", taskInfo.Namespace, taskInfo.Name, ctx.NodeName)
+		if status == Pipelined {
+			msg += " once resource is released"
+		}
+		return PodReasonSchedulable, msg
+	case Pending:
+		if fe := ji.NodesFitErrors[tid]; fe != nil {
+			// Pod is not schedulable
+			return PodReasonUnschedulable, fe.Error()
+		}
+		// Pod is not scheduled yet
+		return PodReasonUndetermined, msg
```

Comment:
And BTW, l acually cannot fully understand the status of `Undetermined`.  Before commit/discard, if this pod is still pending while not allocated or pipelined, should I also consider it a scheduling failure?

Reply:
This pod is still pending, because it is not scheduled yet. From users' perspective, it's a scheduling failure, but this commit want to expose the specific reason or state. Thus, the  `Undetermined` is used, stands for is not scheduled yet.

Reply:
So as you said it's still `PodReasonUnschedulable ` right? If there is no corresponding NodesFitErrors, I think it should be added?

Reply:
It's different from `PodReasonUnschedulable` which is actually scheduled but failed. What we discussed is not scheduled yet.

Reply:
Do you mean that task will skip scheduling?  l think pending tasks under one job will all go through allocate logic. If they are still pending before commit, I think they should be regarded as scheduling fails?

Reply:
- That's correct. `Undetermined` tasks were skipped while scheduling. 
- Please check https://github.com/volcano-sh/volcano/blob/master/pkg/scheduler/actions/allocate/allocate.go#L211

Reply:
good catch 👍 , sorry l missed that.

Reply:
Finally. :P